### PR TITLE
Add custom request API

### LINKED ISSUE #169: Custom request API
Allow consumers of this library to make custom requests. Provide a request method that handles authentication but allows the consumer to specify the URL path, request headers and body. Provide helpers for parsing the result payload.

### PR DESCRIPTION
**Pull request by ubarsaiyan**
Closes #169 
This PR also includes https://github.com/perry-mitchell/webdav-client/pull/168 by @yoh1496 as it is instrumental in sending custom requests that contain a body.
I have included one test that sends a stat request and passes the stat checks.

### CODE REVIEW

File: source/interface/stat.js [hunk 1]
```diff
@@ -47,5 +47,6 @@ function parseStat(result, filename, isDetailed = false) {
 }
 
 module.exports = {
-    getStat
+    getStat,
+    parseStat
```
Comment (perry-mitchell) [inline source/interface/stat.js]: I think this is extra?

Comment (ubarsaiyan) [reply source/interface/stat.js]: I have used parseStat here https://github.com/perry-mitchell/webdav-client/pull/174/files/6c092e915c46ea0754eb498bca6696b10ff124d2#diff-f5ff98f224fcf64e04d02c119dfb0c33R37
Is it a problem exporting this function?

**Comment by ubarsaiyan** [comment]
@perry-mitchell sorry but I don't get where have I used object spreading? Can you please point to it. I'll then fix asap so it can be merged.

**Comment by perry-mitchell** [comment]
@ubarsaiyan You’re right, sorry, it seems that it’s a dependency - hoek. No idea why it’s failing now but I can’t merge this until the tests pass.

**Comment by intel352** [comment]
@perry-mitchell So the last fully passing tests were 3 months ago. 2 months ago, this PR branch and your platform-upgrade branch both started failing on Node 6 and WEB.

It might be worth forcing tests to run against master right now, to make sure it's still fully passing. Perhaps one of your dependencies is no longer compatible with Node 6, so you perhaps need to add a restriction on the dependency version.

That being said, while I'm sure some people use the browser support, how important is it to continue servicing Node 6? Node 6 stopped being supported in May 2019.
Node 8 will stop being supported in Jan 2020.

I'd suggest you at least drop Node 6.

**Comment by intel352** [comment]
@ubarsaiyan Thanks for this branch. Testing this, I would expect customRequest to behave the same as the other requests, in that the base URL should be prepended to a partial URL path.
If you want to support wholly custom URL endpoints in the customRequest, then I'd suggest you should additionally provide a static method for customRequest.

Otherwise, I don't see what benefit is gained from using customRequest on the client instance with the same auth/credentials but the potential to post to a wholly different domain.

File: source/interface/custom.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const { handleResponseCode } = require("../response.js");
+const { prepareRequestOptions, request } = require("../request.js");
+
+function customRequest(requestOptions, options) {
+    prepareRequestOptions(requestOptions, options);
```
Comment (intel352) [inline source/interface/custom.js]: Change to:
```
function customRequest(remotePath, requestOptions, options) {
    if (!requestOptions.url) {
        requestOptions.url = joinURL(options.remoteURL, encodePath(remotePath), "/");
    }
    prepareRequestOptions(requestOptions, options);
```

File: source/factory.js [hunk 3]
```diff
@@ -179,6 +180,28 @@ function createClient(remoteURL, opts = {}) {
...
+         *          headers: {
+         *              Accept: "text/plain",
+         *              Depth: 0
+         *          },
+         *          responseType: "text"
+         *      });
+         */
+        customRequest: function customRequest(requestOptions, options) {
+            const customOptions = merge(runtimeOptions, options || {});
+            return custom.customRequest(requestOptions, customOptions);
```
Comment (intel352) [inline source/factory.js]: Change to:
```
        customRequest: function customRequest(remotePath, requestOptions, options) {
            const customOptions = merge(runtimeOptions, options || {});
            return custom.customRequest(remotePath, requestOptions, customOptions);
```

File: source/factory.js [hunk 4]
```diff
@@ -179,6 +180,28 @@ function createClient(remoteURL, opts = {}) {
...
 
+        /**
+         * Send a custom request
+         * @param {RequestOptions=}  requestOptions the request options
+         * @param {Options=} options Options for the request
+         * @memberof ClientInterface
+         * @returns {Promise<Any>} A promise that resolves with response of the request
+         * @example
+         *      const contents = await client.customRequest({
+         *          url: "http://localhost:9988/webdav/server/alrighty.jpg",
```
Comment (intel352) [inline source/factory.js]: Change to:
```
         *      const contents = await client.customRequest("/alrighty.jpg", {
```

File: test/specs/customRequest.spec.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    afterEach(function() {
+        return this.server.stop();
+    });
+
+    it("send and parse stat custom request", function() {
+        let response = null;
+        return this.client
+            .customRequest({
+                url: "http://localhost:9988/webdav/server/alrighty.jpg",
```
Comment (intel352) [inline test/specs/customRequest.spec.js]: Change to:
```
            .customRequest("/alrighty.jpg", {
```

**Comment by ubarsaiyan** [comment]
@intel352 Thank you for your suggestions. I have made the required changes. cc: @perry-mitchell @guilherme.gazzo

**Comment by perry-mitchell** [comment]
Great, so all that's left is to solve the failing tests.. Node 6:

```
/home/travis/build/perry-mitchell/webdav-client/node_modules/@hapi/joi/lib/types/object/index.js:255
                        !pattern.schema._validate(key, state, { ...options, abortEarly:true }).errors) {
                                                                ^^^
SyntaxError: Unexpected token ...
```

And web:

```
$ export DISPLAY=:99.0
0.00s$ sh -e /etc/init.d/xvfb start
sh: 0: Can't open /etc/init.d/xvfb
The command "sh -e /etc/init.d/xvfb start" failed and exited with 127 during .
```

Looking into this on master now..

**Comment by ubarsaiyan** [comment]
@perry-mitchell yes I think this is not specific to my PR. Thanks!

**Comment by perry-mitchell** [comment]
Node issue related to: jeffbski/wait-on#44

**Comment by perry-mitchell** [comment]
@ubarsaiyan Master should be fixed.. could you please merge?

**Comment by ubarsaiyan** [comment]
@perry-mitchell done

**Comment by perry-mitchell** [comment]
I won’t be home for a bit, but I’ll most likely release later today. Thanks!

**Comment by perry-mitchell** [comment]
Released under 2.10.0! Thanks

**Pull request merged**